Fix(homeserver): DDoS due to slow uploads or downloads

SeverinAlexB opened the pull request.

SHAcollision reviewed (commented):
This PR looks super. I think it's basically ready to merge, but I left a few questions.

Other small suggestions, no action needed:
- Add metrics/logging: if we instrument read/write durations, LMDB transaction durations and error counts, that might help diagnose any remaining hot spots. In this regard, do you think adding opentelemetry to the homeserver might be valuable? We already use it in Nexus.

- Consider streaming "deletes": for very large directories, `delete_file` iterates chunk by chunk in a blocking `txn`. If someone deletes a multi-GB file, that blocking period could be long, well, maybe, I don't know. Is there any way we could apply the same temp‐file trick or shard deletes across multiple txns?

- Prevent someone filling /tmp: now that we’re disk-buffering, we might need to apply per-tenant tmp-disk quotas to prevent someone filling /tmp. E.g., a single user slowly writes thousands of 1GB, eventually exceeding his quota as none of it has yet reached LMDB and updated the `user.used_bytes` . Worth checking if this is possible, if so we can fix on a follow up PR.

I leave here some LLM comments that I can't judge if they deserve any merit but maybe you understand them:

> - `WebDavPath::new` requires `/pub/` (trailing slash) – hence GET /pub (directory listing) will 400. That looks deliberate, just flagging that davfs clients expect the root collection to be listable.
> - Streaming reads still allocate a `Vec<u8>` per LMDB chunk before writing it to the `tokio::fs::File;` zero-copy `bytes::Bytes` + `File::write_all_vectored` would shave some µs.
> - `pub(crate) fn max_chunk_size() -> usize` became `pub fn` in order to use it in tests; consider gating with `#[cfg(test)]` instead so the public API stays minimal.

SeverinAlexB commented on pubky-homeserver/src/persistence/lmdb/tables/files/blobs.rs:
```diff
@@ -0,0 +1,153 @@
+        let db = self.clone();
+        let id = *id;
+        tokio::task::spawn_blocking(move || -> anyhow::Result<InDbTempFile> {
+            db.read_file_sync(&id)
```
Any anyhow error is turned into a HTTP 500 so this should already happen. Or am I missing something?

I did add a log as you suggested to all spawns.

SeverinAlexB commented on pubky-homeserver/src/core/routes/tenants/write.rs:
```diff
@@ -66,36 +68,38 @@ pub async fn delete(
     let hint = body.size_hint().exact();
     if let Some(exact_bytes) = hint {
         enforce_user_disk_quota(existing_entry_bytes, exact_bytes, used_bytes, quota_bytes)?;
     }
 
-    // Stream body
-    let mut writer = state.db.write_entry(public_key, full_path)?;
+    // Stream body to disk first.
     let mut seen_bytes: u64 = 0;
     let mut stream = body.into_data_stream();
+    let mut buffer_file_writer = AsyncInDbTempFileWriter::new().await?;
 
     while let Some(chunk) = stream.next().await.transpose()? {
         seen_bytes += chunk.len() as u64;
         enforce_user_disk_quota(existing_entry_bytes, seen_bytes, used_bytes, quota_bytes)?;
-        writer.write_all(&chunk)?;
+        buffer_file_writer.write_chunk(&chunk).await?;
     }
```
Sure. Not sure this should be part of this PR though? Do you want to add one? :) Obviously, if you are busy either Ilya or I can do it next week.

SeverinAlexB commented on pubky-homeserver/src/shared/webdav/webdav_path.rs:
```diff
@@ -0,0 +1,343 @@
+const MAX_WEBDAV_PATH_TOTAL_LENGTH: usize = 4096;
+
+/// Takes a path, normalizes and validates it.
+/// Make sure to url decode the path before calling this function.
```
> /// Make sure to url decode the path before calling this function.

Isn't it already documented or what do you mean specifically?

SHAcollision replied:
Wops, the `NIT:` part came intuitively :)

SeverinAlexB commented:
Incorporated feedback and added questions where unclear.


> Add metrics/logging: if we instrument read/write durations, LMDB transaction durations and error counts, that might help diagnose any remaining hot spots. In this regard, do you think adding opentelemetry to the homeserver might be valuable? ?We already use it in Nexus.

Can you give me an opentelemetry intro at some point? I wanted to add sentry. Not sure if it is the same or which is superior.

> Consider streaming "deletes": for very large directories, delete_file iterates chunk by chunk in a blocking txn. If someone deletes a multi-GB file, that blocking period could be long, well, maybe, I don't know. Is there any way we could apply the same temp‐file trick or shard deletes across multiple txns?

As already discussed in Slack, we'll not tackling this work as we'll move to a Google Buckets soon.

> Prevent someone filling /tmp: now that we’re disk-buffering, we might need to apply per-tenant tmp-disk quotas to prevent someone filling /tmp. E.g., a single user slowly writes thousands of 1GB, eventually exceeding his quota as none of it has yet reached LMDB and updated the user.used_bytes . Worth checking if this is possible, if so we can fix on a follow up PR.

Same as deletes. This problem will go away as soon as we move to Google Buckets.


> - WebDavPath::new requires /pub/ (trailing slash) – hence GET /pub (directory listing) will 400. That looks deliberate, just flagging that davfs clients expect the root collection to be listable.

You are correct. Hmmm. I need to fix that.

- Streaming reads still allocate a Vec<u8> per LMDB chunk before writing it to the tokio::fs::File; zero-copy bytes::Bytes + File::write_all_vectored would shave some µs.

Will need to rewrite this anyway with Google Buckets. Good catch though :)

- pub(crate) fn max_chunk_size() -> usize became pub fn in order to use it in tests; consider gating with #[cfg(test)] instead so the public API stays minimal.

I don't see it being used in tests but LMDB::open. I removed the the `pub` qualifier.
